controllers/ctltools.js

```diff
@@ -0,0 +1,302 @@
+  req.body.isTemplate = true;
+  if(create || !req.body.isLocal) { // non-local (LTI) tools can be updated forever, local (OAuth2) only during creation
+    req.body.oAuthClientId = req.body.oAuthClientId || "";
+  } else {
+    req.body.oAuthClientId = undefined; // undefine the property prohibits database update
+  }
+  req.body.isLocal = (create ? !!req.body.isLocal : undefined);
+  req.body.scope = req.body.scope || "openid offline";
+  req.body.skipConsent = !!req.body.skipConsent;
+  req.body.openNewTab = !!req.body.openNewTab;
+  req.body.frontchannel_logout_uri = req.body.frontchannel_logout_uri || null;
+  req.body.isHidden = !!req.body.isHidden;
+  return req;
+};
+
+const createTool = (req, service, next) => {
+  api(req).post('/' + service + '/', {
+    json: req.body
+  }).then(tool => {
+    next();
+  }).catch(err => {
+    if(req.body.isLocal) {
+      api(req, { version: HYDRA_VERSION }).delete(`/oauth2/clients/${req.body.oAuthClientId}`).then(_ => {
+        next(err);
+      });
+    }
+    next(err);
+  });
+};
+
+const getCreateHandler = (service) => {
+    return function (req, res, next) {
+      req = sanitizeTool(req, true);
+      if(req.body.isLocal) {
+        return api(req, { version: HYDRA_VERSION }).post('/oauth2/clients/', {
+          json: getClient(req.body, true)
+        }).then(response => {
+          req.body.oAuthClientId = response.client_id;
+          createTool(req, service, next);
+        }).catch(err => {
+          next(err);
+        });
+      } else {
+        createTool(req, service, next);
+      }
+    };
+};
+
+const getUpdateHandler = (service) => {
+    return function (req, res, next) {
+      req = sanitizeTool(req);
+      return api(req).patch('/' + service + '/' + req.params.id, {
+          json: req.body
+      }).then(data => {
+        if(data.isLocal) {
+          return api(req, { version: HYDRA_VERSION }).put(`/oauth2/clients/${data.oAuthClientId}`, {
+            json: getClient(req.body)
+          }).then(_ => {
+            res.redirect(req.header('Referer'));
+          });
+        }
+      res.redirect(req.header('Referer'));
+      }).catch(err => {
+          next(err);
+      });
+    };
+};
+
+const getDetailHandler = (service) => {
+    return function (req, res, next) {
+        api(req).get('/' + service + '/' + req.params.id).then(data => {
+          if(data.isLocal) {
+            api(req, { version: HYDRA_VERSION }).get(`/oauth2/clients/${data.oAuthClientId}`).then(client => {
+              data.secret = PASSWORD;
+              data.redirect_url = client.redirect_uris.join(";");
+              data.token_endpoint_auth_method = client.token_endpoint_auth_method;
+              data.scope = client.scope;
+              data.frontchannel_logout_uri = client.frontchannel_logout_uri;
+              res.json(data);
+            });
+          } else {
+            data.secret = PASSWORD;
```

Comment:
I can't understand why secret equals "******" in two places ?

Reply:
the controllers/ctltools.js is mainly copied from the controllers/tools.js. We are doing necessary changes to it in https://github.com/hpi-schul-cloud/superhero-dashboard/pull/184